### PR TITLE
Implement HTTP Message Signatures

### PR DESCRIPTION
**Pull request by TillaTheHun0**

### CODE REVIEW

File: src/hb_http_signature.erl [hunk 1]
```diff
@@ -1,10 +1,44 @@
...
+	status => integer(),
+	headers => fields(),
+	trailers => fields()
+}.
+-type component_identifier() :: {
+	item,
+	{string, binary()},
+	{binary(), integer() | boolean() | {string | token | binary, binary()}}
+}.
+-type signature_params() :: #{binary() => binary() | integer()}.
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl]: I added additional comments with examples

File: src/hb_http_signature.erl [hunk 2]
```diff
@@ -1,10 +1,44 @@
...
+-type authority_state() :: #{
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:66]: I added additional comments with examples

File: src/hb_http_signature.erl [hunk 3]
```diff
@@ -13,68 +47,86 @@
 %%% verify(Authority, SigName, Msg) -> {ok}
 %%%
 
+-spec authority(binary(), #{binary() => binary() | integer()}, binary()) -> authority_state().
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl]: Yep, good catch. I've updated.

File: src/hb_http_signature.erl [hunk 4]
```diff
@@ -13,68 +47,86 @@
...
 authority(ComponentIdentifiers, SigParams, Key) ->
 	#{
-		component_identifiers => ComponentIdentifiers,
+		% Since parsing is performed here, this provides a feedback loop
+		% for properly shaped component identifiers
+		component_identifiers => lists:map(fun sf_item/1, ComponentIdentifiers),
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:102]: I've add comment to clarify. The interface supports both binary encoded and parsed Structured Field Items to be passed in. In the case of encoded values, they must be parsed. In the case of parsed, they must be validated. `sf_item/1` handles both of those cases via header matching

File: src/hb_http_signature.erl [hunk 5]
```diff
@@ -175,66 +265,68 @@ extract_field_value(RawFields, [Key, IsStrictFormat, IsByteSequenceEncoded]) ->
...
 					<<"@method">> ->
-						{ok, {NormalizedItem, upper_bin(maps:get(method, Req))}};
+						{ok, upper_bin(maps:get(method, Req))};
 					% https://datatracker.ietf.org/doc/html/rfc9421#section-2.2-4.4.1
 					<<"@target-uri">> ->
-						{ok, {NormalizedItem, bin(maps:get(url, Req))}};
+						{ok, bin(maps:get(url, Req))};
 					% https://datatracker.ietf.org/doc/html/rfc9421#section-2.2-4.6.1
 					<<"@authority">> ->
 						URI = uri_string:parse(maps:get(url, Req)),
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:363]: The url isn't needed for all derived components, and so derived components need to parse url later, for example after validation of other parameters. In those cases, the url may never need to be parsed ie. a parameter validation error occurs.

Finally, purely from a style perspective, I like that each branch of the `case` statement is self-contained and there is no shared data between them (besides of course the function arguments). In other words, I could delete/re-write an entire clause and not have to worry about affecting other clauses. Personally, I like this better for maintainability.

File: src/hb_http_signature.erl [hunk 6]
```diff
@@ -381,3 +518,331 @@ random_an_binary(Length) ->
 	RandomIndexes = [rand:uniform(ListLength) || _ <- lists:seq(1, Length)],
 	RandomChars = [lists:nth(Index, Characters) || Index <- RandomIndexes],
 	list_to_binary(RandomChars).
+
+trim_ws(<<$\s, Bin/bits>>) -> trim_ws(Bin);
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:564]: Yep. I've added some comments and a test. Basically, it trims the whitespace from the front and end of the binary.

File: src/hb_http_signature.erl [hunk 7]
```diff
@@ -87,23 +139,46 @@ signature_params_line(ComponentIdentifiers, SigParams) when is_list(SigParams) -
...
 
-extract_field(Identifier, Req, Res = #{}) ->
+%%% @doc Given a Component Identifier and a Request/Response Messages Context
+%%% extract the value represented by the Component Identifier, from the Messages Context,
+%%% specifically a field on a Message within the Messages Context,
+%%% and return the normalized form of the identifier, along with the extracted encoded value.
+%%%
+%%% This implements a portion of RFC-9421
+%%% See https://datatracker.ietf.org/doc/html/rfc9421#name-http-fields
+extract_field(Identifier, Req, Res) when map_size(Res) == 0 ->
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:218]: Yep, good catch, i've removed.

File: src/hb_http_signature.erl [hunk 8]
```diff
@@ -122,12 +197,17 @@ extract_field(Identifier, Req, Res, _Subject) ->
...
+							end,
 							% The header may exist on any message in the context of the signature
 							% which could be the Request or Response Message
 							% https://datatracker.ietf.org/doc/html/rfc9421#section-2.1-18.8.1
-							IsRequestIdentifier andalso Req orelse Res,
-							#{}
+							case IsRequestIdentifier of
+								true -> Req;
+								false -> Res
+							end#{}
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl]: Good catch, i've removed.

File: src/hb_http_signature.erl [hunk 9]
```diff
@@ -122,12 +197,17 @@ extract_field(Identifier, Req, Res, _Subject) ->
 						maps:get(
 							% The field will almost certainly be a header, but could also be a trailer
 							% https://datatracker.ietf.org/doc/html/rfc9421#section-2.1-18.10.1
-							IsTrailerField andalso trailers orelse headers,
+							case IsTrailerField of
```
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl]: I've updated the comment to reflect why I did it this way.

In the RFC-9421, the message field names included in the signature base are normalized to lowercase. In HTTP RFC, message field name equality is also case-insensitive. So by converting all the names to lowercase here, we simultaneously normalize them for RFC-9421, and prepare them for comparison according to HTTP Semantics, all in one pass.

File: src/hb_http_signature.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (oltarasenko) [reply src/hb_http_signature.erl:102]: Do we have to support two interfaces at the same time?

File: src/hb_http_signature.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (oltarasenko) [reply src/hb_http_signature.erl:363]: ok, makes sense!

File: src/hb_http_signature.erl (hunk 4 above)
[the remark this answers is not in the document]
Comment (TillaTheHun0) [reply src/hb_http_signature.erl:102]: I suppose not, but it makes the code a lot simpler, and more declarative.

`sf_item` is used internally as well, during parsing of HTTP fields as well, which may already be parsed themselves or need to be parsed.

File: src/hb_http_signature.erl [hunk 10]
```diff
@@ -381,3 +559,342 @@ random_an_binary(Length) ->
...
+			<<Trimmed:S/binary, _/bits>> = Value,
+			Trimmed
+	end.
+
+%%%
+%%% TESTS
+%%%
+
+trim_ws_test() ->
+	<<"hello world">> = trim_ws(<<"      hello world      ">>),
```
Comment (oltarasenko) [inline src/hb_http_signature.erl:588]: I would suggest using the ?assertEqual, as the usual way of verifying with erlang unit tests. For example the output from match verification is:
```
in function hb_http_signature:trim_ws_test/0 (/Users/oleg/repos/arweave/HyperBEAM/src/hb_http_signature.erl, line 588)
in call from eunit_test:'-mf_wrapper/2-fun-0-'/2 (eunit_test.erl, line 274)
in call from eunit_test:run_testfun/1 (eunit_test.erl, line 72)
in call from eunit_proc:run_test/1 (eunit_proc.erl, line 544)
in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 369)
in call from eunit_proc:handle_test/2 (eunit_proc.erl, line 527)
in call from eunit_proc:tests_inorder/3 (eunit_proc.erl, line 469)
in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 359)
**error:{badmatch,<<"hello world     \t">>}
  output:<<"">>
```
(aka, some badmatch on line 588 :))

With ?assertEqual you have:
```
hb_http_signature: trim_ws_test...*failed*
in function hb_http_signature:trim_ws_test/0 (/Users/oleg/repos/arweave/HyperBEAM/src/hb_http_signature.erl, line 588)
in call from eunit_test:'-mf_wrapper/2-fun-0-'/2 (eunit_test.erl, line 274)
in call from eunit_test:run_testfun/1 (eunit_test.erl, line 72)
in call from eunit_proc:run_test/1 (eunit_proc.erl, line 544)
in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 369)
in call from eunit_proc:handle_test/2 (eunit_proc.erl, line 527)
in call from eunit_proc:tests_inorder/3 (eunit_proc.erl, line 469)
in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 359)
**error:{assertEqual,[{module,hb_http_signature},
              {line,588},
              {expression,"trim_ws ( << \"      hello world    \" >> )"},
              {expected,<<"hello world  !">>},
              {value,<<"hello world">>}]}
  output:<<"">>
```

so: We tried to call trim_ws ( << \"      hello world    \" >> ), got <<"hello world">> expected: <<"hello world  !">>

Comment (TillaTheHun0) [reply src/hb_http_signature.erl]: I've moved all the tests to use the `assertEqual` macro

**Pull request merged**